unyt/tests/test_array_functions.py

```diff
@@ -50,14 +50,47 @@ def test_dot_matrices(out):
 
     if out is not None:
         np.testing.assert_array_equal(res, out)
-        assert res is out
+        assert np.shares_memory(res, out)
+        assert isinstance(res, unyt_array)
+        assert isinstance(out, np.ndarray)
 
     if isinstance(out, unyt_array):
         # check that the result can be converted to predictible units
         res.in_units("cm * s")
         assert out.units == res.units
 
 
+def test_dot_mixed_ndarray_unyt_array():
+    a = np.ones((3, 3))
+    b = np.ones((3, 3)) * cm
+
+    res = np.dot(a, b)
+
+    assert isinstance(res, unyt_array)
+    assert res.units == cm
+
+    out = np.zeros((3, 3))
+    res = np.dot(a, b, out=out)
+
+    assert isinstance(res, unyt_array)
+    assert type(out) is np.ndarray
+    np.testing.assert_array_equal(out, res)
```

Comment:
I don't know *if* we can pass this too, but it'd be great to check that memory usage isn't doubled
(this is debatable but at the moment I believe it'd be the preferable behaviour, provided it's possible to implement. In any case this is the kind of subtle gotcha we'll need to document)
```suggestion
    np.testing.assert_array_equal(out, res)
    assert out.base is res
```

Reply:
resolved with `numpynp.shares_memory`